libcudacxx/include/cuda/std/__algorithm/sort.h

```diff
@@ -0,0 +1,1129 @@
+
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
+
+// stable, 2-3 compares, 0-2 swaps
+
+template <class _AlgPolicy, class _Compare, class _ForwardIterator>
+_LIBCUDACXX_HIDE_FROM_ABI constexpr unsigned
+__sort3(_ForwardIterator __x, _ForwardIterator __y, _ForwardIterator __z, _Compare __c)
+{
+  using _Ops = _IterOps<_AlgPolicy>;
+
+  unsigned __r = 0;
+  if (!__c(*__y, *__x)) // if x <= y
+  {
+    if (!__c(*__z, *__y)) // if y <= z
+    {
+      return __r; // x <= y && y <= z
+                  // x <= y && y > z
+    }
+    _Ops::iter_swap(__y, __z); // x <= z && y < z
+    __r = 1;
+    if (__c(*__y, *__x)) // if x > y
+    {
+      _Ops::iter_swap(__x, __y); // x < y && y <= z
+      __r = 2;
+    }
+    return __r; // x <= y && y < z
+  }
+  if (__c(*__z, *__y)) // x > y, if y > z
+  {
+    _Ops::iter_swap(__x, __z); // x < y && y < z
+    __r = 1;
+    return __r;
+  }
+  _Ops::iter_swap(__x, __y); // x > y && y <= z
+  __r = 1; // x < y && x <= z
+  if (__c(*__z, *__y)) // if y > z
+  {
+    _Ops::iter_swap(__y, __z); // x <= y && y < z
+    __r = 2;
+  }
+  return __r;
+} // x <= y && y <= z
+
+// stable, 3-6 compares, 0-5 swaps
+
+template <class _AlgPolicy, class _Compare, class _ForwardIterator>
+_LIBCUDACXX_HIDE_FROM_ABI void
+__sort4(_ForwardIterator __x1, _ForwardIterator __x2, _ForwardIterator __x3, _ForwardIterator __x4, _Compare __c)
+{
+  using _Ops = _IterOps<_AlgPolicy>;
+  _CUDA_VSTD::__sort3<_AlgPolicy, _Compare>(__x1, __x2, __x3, __c);
+  if (__c(*__x4, *__x3))
+  {
+    _Ops::iter_swap(__x3, __x4);
+    if (__c(*__x3, *__x2))
+    {
+      _Ops::iter_swap(__x2, __x3);
+      if (__c(*__x2, *__x1))
+      {
+        _Ops::iter_swap(__x1, __x2);
+      }
+    }
+  }
+}
+
+// stable, 4-10 compares, 0-9 swaps
+
+template <class _AlgPolicy, class _Comp, class _ForwardIterator>
+_LIBCUDACXX_HIDE_FROM_ABI void __sort5(
+  _ForwardIterator __x1,
+  _ForwardIterator __x2,
+  _ForwardIterator __x3,
+  _ForwardIterator __x4,
+  _ForwardIterator __x5,
+  _Comp __comp)
+{
+  using _Ops = _IterOps<_AlgPolicy>;
+
+  _CUDA_VSTD::__sort4<_AlgPolicy, _Comp>(__x1, __x2, __x3, __x4, __comp);
+  if (__comp(*__x5, *__x4))
+  {
+    _Ops::iter_swap(__x4, __x5);
+    if (__comp(*__x4, *__x3))
+    {
+      _Ops::iter_swap(__x3, __x4);
+      if (__comp(*__x3, *__x2))
+      {
+        _Ops::iter_swap(__x2, __x3);
+        if (__comp(*__x2, *__x1))
+        {
+          _Ops::iter_swap(__x1, __x2);
+        }
+      }
+    }
+  }
+}
+
+// The comparator being simple is a prerequisite for using the branchless optimization.
+template <class _Tp>
+struct __is_simple_comparator : false_type
+{};
+template <>
+struct __is_simple_comparator<__less&> : true_type
+{};
+template <class _Tp>
+struct __is_simple_comparator<less<_Tp>&> : true_type
+{};
+template <class _Tp>
+struct __is_simple_comparator<greater<_Tp>&> : true_type
+{};
+
+template <class _Compare, class _Iter, class _Tp = typename iterator_traits<_Iter>::value_type>
+using __use_branchless_sort =
```

Comment:
Q: Would we want to use a variable template and `if constexpr` below instead of tag dispatch?